packages/riverpod/test/devtool_test.dart

```diff
@@ -0,0 +1,218 @@
+// ignore_for_file: avoid_types_on_closure_parameters
+
+import 'package:riverpod/src/internals.dart';
+import 'package:test/test.dart';
+
+void main() {
+  tearDown(() {
+    RiverpodBinding.debugInstance.containers = {};
+    stopSpyPostEvent();
+  });
+
+  test('should add container to containers map', () {
+    final firstContainer = ProviderContainer();
+    final secondContainer = ProviderContainer();
+
+    final container = RiverpodBinding.debugInstance.containers;
+
+    expect(
+      container,
+      equals({
+        firstContainer.debugId: firstContainer,
+        secondContainer.debugId: secondContainer,
+      }),
+    );
+  });
+
+  test('should map containers to ContainerNode', () {
+    final firstContainer = ProviderContainer();
+    final secondContainer = ProviderContainer();
+    final stringProvider = Provider(
+      (_) => 'some string',
+      name: 'stringProvider',
+    );
+    final intProvider = Provider((_) => 10);
+    firstContainer.read(stringProvider);
+    secondContainer.read(intProvider);
+    final stringProviderId = firstContainer.debugIdForProvider(stringProvider);
+    final intProviderId = secondContainer.debugIdForProvider(intProvider);
+
+    final containerNodes = RiverpodBinding.debugInstance.containerNodes;
+
+    expect(
+      containerNodes,
+      predicate(
```

Comment:
well, this is a creative solution to avoid adding `Equatable` or `Freezed` to the package. 

AS I can't compara two objects directly I created the extension functions to make the fields check without the need of changing production code. It is not the prettiest code, I know. Do you have any suggestions?

Reply:
Feel free to override ==/hashCode for those classes.

Reply:
perfect! This will be way cleaner hehe

Reply:
added here https://github.com/rrousselGit/riverpod/pull/1471/commits/3fe50950a8dfef108882fa5724977ecd23e62ccd